Client isolation announcement (#3549)
SoftwareAG/c8y-docs: content/change-logs/platform-services/mqtt-service-0.9.x-device-isolation.md

Reviewer: As a customer who needs inter-device communication, I'm a bit worried and need more explicit reassurance that it will still be possible to write an application (microservice) that lets me broker inter-device communication. I can infer it from the text and my own knowledge, but it ought to be in black and white.

Author: I can re-word it, but that point was supposed to be captured by line 32:
> These changes will be announced soon, and will allow messages to be sent to other clients, under the control of the application microservice.

Actually I think @jato-c8y's [proposed re-org](https://github.com/Cumulocity-IoT/c8y-docs/pull/3549#discussion_r2207132108) will address this in a much nicer way. Stay tuned.

Reviewer: By the time I get to that line, my blood pressure is already raised because you buried the lede :-)

Two suggestions:
* In that line, make explicit that you can write some kind of 'broker' that relays messages based on your custom needs
* "but it will **not** receive messages published by any other client" .... "unless explicitly routed by an application"

Author: Fair! I guess I'm struggling a bit to keep the discussion focused on the isolation changes without also explaining everything that will come later with the topic mapping and Pulsar client. Points noted though and I will try to include this in my second draft.

---
date: 2025-07-21
title: MQTT Service will enforce device-level isolation
change_type:
  - value: change-3BQrQ6adS
    label: API change
product_area: Platform services
component:
  - value: component-LcWEQW5gs
    label: MQTT
build_artifact:
  - value: tc-hc5Tfixeqqei
    label: mqtt-service
issue: MTM-64100
---

The Public Preview release of the {{< product-c8y-iot >}} [MQTT Service](/device-integration/mqtt-service/) currently enforces *tenant-level* isolation.

> An MQTT client connected to the service can subscribe to any topic, and receive messages published by other clients connected to the same tenant.

Reviewer: ```suggestion
An MQTT client connected to the service can subscribe to any topic, and receive messages published by other clients to that topic on the same tenant.
```
Bit nitpicky, but somebody in a hurry could misread this.

Author: https://github.com/Cumulocity-IoT/c8y-docs/pull/3549/commits/6df7d851b0ea6015b5895accb552379fb9e333cd

> An MQTT client connected to the service can subscribe to any topic, and receive messages published by other clients connected to the same tenant.

Reviewer 2: suggestion (mainly agreeing with existing comment made and a minor opinion):
I agree with comment you made in summary about device vs client.
In this opening remark I think you should state 'An MQTT client (device)' and then stick to client or device thereafter.
I also think 'the service' is the first mention of the MQTT Service excluding the link above so should be named explicitly.

Author: https://github.com/Cumulocity-IoT/c8y-docs/pull/3549/commits/6df7d851b0ea6015b5895accb552379fb9e333cd (use "MQTT Service" everywhere)

On the device vs. client question I'm going to get that resolved once and for all in the steering meeting tomorrow. We all say "device" without really thinking about it, but the connected client _could_ be handling messages for many devices.

For the upcoming Generally Available (GA) release, the MQTT Service will enforce *device-level* isolation.
A client (device) will still be able to subscribe to any topic, but it will **not** receive messages published by any other client.
In effect, each client will have its own private topic space that is not shared with other clients.

We are making this change to align the MQTT Service with the behavior of the existing [Core MQTT](/device-integration/mqtt/) capability, and to improve the out-of-the-box security for typical IoT applications where inter-device communication is not required.

This is a **breaking change** and affected applications **must** be updated to continue working after the GA release of the MQTT Service.
The change will affect any applications where MQTT clients communicate by publishing and subscribing to the same topics.
In particular it will affect applications where a {{< product-c8y-iot >}} microservice connects to the MQTT Service using MQTT rather than the client SDK, to receive messages from or send messages to MQTT devices.

Please note that breaking changes to the MQTT Service client SDK are also planned for the GA release.

> These changes will be announced soon, and will allow messages to be sent to other clients, under the control of the application microservice.

Reviewer 2: ```suggestion
Details of the required changes will be announced soon, and will allow messages to be sent to other clients, under the control of the application microservice.
```

suggestion (just some opinions):
I have changed the opening here because I am not sure it is clear if this sentence belongs in this paragraph or applies to the whole announcement.
I also think it probably puts too much emphasis on still allowing messages to be sent to other clients, but in the opening remark we emphasise 'it will not receive messages published by any other client'.
'and will allow messages to be sent to other clients, under the control of the application microservice.' -> 'there will be the possibility, if explicitly desired, for an application microservice to determine and be used to send messages to other clients' (bit wordy, but maybe needs some clarity?)

Author: https://github.com/Cumulocity-IoT/c8y-docs/pull/3549/commits/6df7d851b0ea6015b5895accb552379fb9e333cd

> Applications that will be affected by device-level isolation, and that are **not** currently using the client SDK, should wait for these changes to be announced before starting their migration activity.

Reviewer 2: suggestion (I see some potential confusion): 

I don't follow these 'Applications' lines.
Should they be in their own paragraph?

Or 'these changes are announced' is actually now referring to just the SDK changes?

I guess what I feel perhaps would help me would be headings for personas?
### General update

...
Exact changes required to ensure device isolation will work will be announced soon. 
...

### If you are currently using the client SDK
Please note that breaking changes to the MQTT Service client SDK are also planned for the GA release.
...

Author: I like this idea. I will re-work it along these lines before I open it up for a proper review.

When the MQTT Service transitions from Public Preview to General Availability (GA), the MQTT Service will enforce *device* isolation.
An MQTT device will still be able to subscribe to any topic, but it will not **automatically** receive messages published to that topic by any other MQTT device.
However, microservices will be able to explicitly route messages between different devices.
In effect, each MQTT client identifier will have its own private topic space that is not shared with other clients, but can be accessed by microservices.

We are making this change to align the MQTT Service with the behavior of the existing [Core MQTT](/device-integration/mqtt/) capability, and to improve out-of-the-box security for typical IoT applications where direct inter-device communication is not required.

### Impact on MQTT devices and microservices connecting to the MQTT Service

This is a **breaking change** and affected applications **must** be updated to continue working after the GA version of the MQTT Service is deployed.

This change affects any application where MQTT devices exchange messages by publishing and subscribing to the same topics.

> In particular it affects applications where a {{< product-c8y-iot >}} microservice connects to the MQTT Service using the MQTT protocol rather than the {{< product-c8y-iot >}} client SDK, to exchange messages with connected MQTT devices.

Reviewer: ```suggestion
In particular it affects applications where a {{< product-c8y-iot >}} microservice connects to the MQTT Service using the MQTT protocol to exchange messages with connected MQTT devices.
```
This part of the sentence is dangling the idea that the right way to do it is the SDK, and then the next line cruelly rips it away. I think it confuses more than it adds information.

Author: https://github.com/Cumulocity-IoT/c8y-docs/pull/3549/commits/51e539232ec47333a1d53428aa7c3adeee73872d

Microservices that are already using the Java client SDK may continue to do so until the new API is available.

### Roll-out plan

To allow customers to plan a managed migration to the new device isolation behavior, it will be rolled out in several phases, as described below.

Dates refer to when the change will reach the `eu-latest` environment; other environments will be updated later following the usual CD deployment schedule.

#### Phase 1: Introduction of device isolation
On or soon after August 4, 2025:
* Tenants currently using the MQTT Service will continue to have tenant isolation by default.
* Tenants **not** currently using the MQTT Service will have device isolation enabled if they later start to use the MQTT Service.
* A public feature toggle will be introduced, allowing tenants to be switched between tenant and device isolation under user control.

#### Phase 2: Migration during Public Preview
* Tenants starting to use the MQTT Service for the first time should develop their applications to work with device isolation.
* Tenants already using the MQTT Service should update their applications to work with device isolation.

> * The feature toggle will allow these tenants to switch between isolation modes while developing and testing changes to their applications.

Reviewer: Logically, isn't this part of the 2nd bullet point rather than a separate issue? That is, you pick one of
* Tenants starting to use ...  (and should not even think about the feature toggle!)
* Tenants already using .... should update ... using the feature toggle

Author: https://github.com/Cumulocity-IoT/c8y-docs/pull/3549/commits/51e539232ec47333a1d53428aa7c3adeee73872d
